src/execution/operator/set/physical_recursive_key_cte.cpp

```diff
@@ -0,0 +1,156 @@
+	DataChunk payload_rows;
+	payload_rows.Initialize(Allocator::DefaultAllocator(), payload_types);
+	PopulateChunk(payload_rows, chunk, payload_idx, true);
+
+	// Add the chunk to the hash table and append it to the intermediate table
+	gstate.ht->AddChunk(distinct_rows, payload_rows, AggregateType::NON_DISTINCT);
+	gstate.intermediate_table.Append(chunk);
+	return SinkResultType::NEED_MORE_INPUT;
+}
+
+//===--------------------------------------------------------------------===//
+// Source
+//===--------------------------------------------------------------------===//
+
+SourceResultType PhysicalRecursiveKeyCTE::GetData(ExecutionContext &context, DataChunk &chunk,
+                                                  OperatorSourceInput &input) const {
+
+	auto &gstate = sink_state->Cast<RecursiveKeyCTEState>();
+	if (!gstate.initialized) {
+		recurring_table->InitializeScan(gstate.scan_state);
+		gstate.finished_scan = false;
+		gstate.initialized = true;
+	}
+	while (chunk.size() == 0) {
+		if (!gstate.finished_scan) {
+			if (chunk.size() == 0) {
+				gstate.finished_scan = true;
+			} else {
+				break;
+			}
+		} else {
+			// we have run out of chunks
+			// now we need to recurse
+			// we set up the working table as the data we gathered in this iteration of the recursion
+
+			// After an iteration, we reset the recurring table
+			// and fill it up with the new hash table rows for the next iteration.
+			if (gstate.intermediate_table.Count() != 0) {
+				recurring_table->Reset();
+				// Set the size of the DataChunks to the maximum of the hash table size and the standard vector size.
+				idx_t size = std::max<idx_t>(gstate.ht->Count(), STANDARD_VECTOR_SIZE);
+				// Initialise the DataChunks to read the resulting rows.
+				// One DataChunk for the payload, one for the keys.
+				DataChunk payload_rows;
+				DataChunk distinct_rows;
+				distinct_rows.Initialize(Allocator::DefaultAllocator(), distinct_types, size);
+				payload_rows.Initialize(Allocator::DefaultAllocator(), payload_types, size);
+
+				// Collect all currently available keys and their payload.
+				gstate.ht->FetchAll(distinct_rows, payload_rows);
+
+				// Create a new DataChunk to store the result.
+				DataChunk result;
+				result.Initialize(Allocator::DefaultAllocator(), chunk.GetTypes(), size);
+				// Populate the result DataChunk with the keys and the payload.
+				PopulateChunk(result, payload_rows, payload_idx, false);
+				PopulateChunk(result, distinct_rows, distinct_idx, false);
+				// Append the result to the recurring table.
+				recurring_table->Append(result);
+			}
+
+			// filling working table
+			working_table->Reset();
+			working_table->Combine(gstate.intermediate_table);
+			gstate.finished_scan = false;
+
+			// and we clear the intermediate table
+			gstate.intermediate_table.Reset();
+			// now we need to re-execute all of the pipelines that depend on the recursion
+			ExecuteRecursivePipelines(context);
```

Comment:
Completely unrelated to this PR - but given you mentioned recursive CTE performance - I've profiled recursive CTEs before and in many cases (almost) all of the time went into the breaking down/setting up of pipeline states. If you're interested in optimizing recursive CTEs, reusing these states across iterations would likely lead to significant performance improvements as well.

Reply:
Oh, that sounds interesting. I will look into it after this Pull-Request.